Create a Table `Row` Type and expose as a Vector on In-Memory Table with `.rows` property

radeusgd opened the pull request:
### Pull Request Description

Implements https://www.pivotaltracker.com/story/show/182307026

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

radeusgd committed:
Implement a basic row type

radeusgd committed:
add some tests

radeusgd committed:
fix

radeusgd committed:
Prototype: Atoms exposing at+length being treated as an array

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java:
```diff
@@ -141,6 +143,48 @@ public Object readMember(String member) throws UnknownIdentifierException {
+  @ExportMessage
+  public boolean hasArrayElements() {
+    Map<String, ?> members = constructor.getDefinitionScope().getMethods().get(constructor.getType());
+    return members != null && members.containsKey("at") && members.containsKey("length");
```
This operation has to be executable on _fast path_ - e.g. with Truffle partial evaluation, otherwise benchmarks are going to be affected.

However PE mode doesn't work well with `Map` - usually all operations with `Map` have to be behind `@TruffleBoundary` (btw. the `engine-runner-native/buildNativeImage` check should be falling with this change). However `@TruffleBoundary` would be too slow...

radeusgd replied:
Yeah, I'm sure that this prototype needs improvement before it can actually be used, I did the simplest possible implementation to see if it even would work.

JaroslavTulach reviewed (changes requested):
The implementation would have to be made ready to run on _fast path_. Anyway, I am not very excited to see this [duck typing](https://discord.com/channels/401396655599124480/1034168034136445028/1034366712008687677) idea overall...

radeusgd committed:
Revert "Prototype: Atoms exposing at+length being treated as an array"

This reverts commit ee05a4a20d45d1f047d38b017cee4ec5079f59cf.

JaroslavTulach committed:
Benchmark to measure 6x slowdown when using ProxyList via host interop

radeusgd committed:
Remove an obsolete error class

radeusgd committed:
Initial implementation of the Array_Proxy

radeusgd committed:
Use Array_Proxy in Table

radeusgd committed:
docs

radeusgd committed:
changelog

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArrayProxy.java:
```diff
@@ -0,0 +1,57 @@
+@ExportLibrary(InteropLibrary.class)
+@ExportLibrary(TypesLibrary.class)
+@Builtin(pkg = "immutable", stdlibName = "Standard.Base.Data.Array_Proxy.Array_Proxy")
+public class ArrayProxy implements TruffleObject {
```
I'd made the class `final` to prevent subclassing. Subclassing doesn't work well with `@ExportMessage` anyway.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArrayProxy.java:
```diff
@@ -0,0 +1,57 @@
+package org.enso.interpreter.runtime.data;
+
+import com.oracle.truffle.api.interop.*;
```
Can we avoid using `*` imports? I'd rather see all the imported types enumerated - in my opinion that helps clarity of the codebase. Especially when reading the code without an IDE.

radeusgd replied:
Sure, I think the IDE added it automatically and I just went with it, but I'm now converting it into single class imports.

JaroslavTulach reviewed (approved):
I believe the `ArrayProxy` implementation is correct. I suggest to expand the `VectorBenchmarks` class with `ArrayProxy` case and compare it compiles well. Use [ProxyList benchmark](https://github.com/enso-org/enso/compare/develop...JaroslavTulach:enso:jtulach/ProxyListExample) as an inspiration.

jdunkerley commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArrayProxy.java:
```diff
@@ -0,0 +1,57 @@
+  private final Object at;
+
+  @Builtin.Method(description = "Creates an array backed by a proxy object.")
+  public ArrayProxy(long length, Object at) {
```
Shouldn't `at` be a `IntFunction<Value>` or similar?

JaroslavTulach replied:
No, it cannot be `IntFunction<Value>`. Such code would be appropriate in _host interop Java_, however here we are writing  a builtin and it needs to be written according to Truffle API - e.g. using `InteropLibrary` rather than `Value`.

jdunkerley replied:
ah ok - thanks for explaining.

radeusgd replied:
Although I think @jdunkerley you have raised a fair point that it is good to check in the constructor that the `Object at` is really an executable function.

I'll add
```
if (CompilerDirectives.inInterpreter()) {
      if (!InteropLibrary.getUncached().isExecutable(at)) {
        throw new IllegalStateException("Array_Proxy needs an executable callback, but got: " + at);
      }
    }
```

inspired by how Vector checks its `storage`.

jdunkerley reviewed (approved):
libs part looks good - happy for this to go in once @JaroslavTulach and @hubertp happy with the built-in part.

radeusgd committed:
cr

radeusgd committed:
Add Array_Proxy benchmark

radeusgd committed:
fixes

radeusgd commented:
I added the benchmarks as requested, for comparison I added the AbstractList benchmark @JaroslavTulach created too.

The results on my machine are following:

<table>
<tr>
    <th> Benchmark
    <th> Time [μs/op]
<tr>
	<td> averageAbstractList
	<td> 119.85
<tr>
	<th> averageOverArrayProxy
	<td> 46.87
<tr>
	<td> averageOverArray
	<td> 18.24
<tr>
	<td> averageOverPolyglotArray
	<td> 18.05
<tr>
	<td> averageOverPolyglotVector
	<td> 16.74
<tr>
	<td> averageOverSlice
	<td> 17.25
</table>

So the proxy is roughly 2.5x slower than direct access (probably due to the additional layer of indirection), but still almost 2.6x faster than the polyglot AbstractList.

JaroslavTulach commented:
> averageOverArrayProxy 	46.87
> averageOverArray 	18.24
> So the proxy is roughly 2.5x slower than direct access

That is very slow. It should be on par with direct access. I'd try the benchmark without the bounds check. E.g. after removing:
```
    if (index >= length || index < 0) {
      throw InvalidArrayIndexException.create(index);
    }
```

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/ArrayProxy.java:
```diff
@@ -0,0 +1,74 @@
+  @ExportMessage
+  public Object readArrayElement(long index, @CachedLibrary(limit = "3") InteropLibrary interop)
+      throws UnsupportedMessageException, InvalidArrayIndexException {
+    if (index >= length || index < 0) {
```
I don't think that's correct or even necessary? If it is outside of the range but you tried to access it you should get UnsupportedMessageException?  Right @JaroslavTulach ?

radeusgd replied:
It is necessary to get the semantics that one would normally expect.

If I remove it, the following test:
```
        Test.specify "should correctly delegate to the callback" <|
            arr = Array_Proxy.new 3 (ix -> ix + 10)
            arr.length . should_equal 3
            arr.at 0 . should_equal 10
            arr.at 1 . should_equal 11
            arr.at 2 . should_equal 12
            arr.at 3 . should_fail_with Index_Out_Of_Bounds_Error_Data
```
will fail like so:
```
Array_Proxy:  [168ms]
    - [FAILED] should correctly delegate to the callback [103ms]
        Reason: Expected an error Constructor<Index_Out_Of_Bounds_Error_Data> but no error occurred, instead got: 13 (at /home/radeusgd/NBO/enso/test/Tests/src/Data/Array_Proxy_Spec.enso:20:13-70).
```

hubertp reviewed (commented):
nit regarding range check
